helios-client/src/main/java/com/spotify/helios/auth/CrtAuthProvider.java

```diff
@@ -0,0 +1,165 @@
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+
+public class CrtAuthProvider implements AuthProvider {
+
+  private static final Logger log = LoggerFactory.getLogger(AuthProviderSelector.class);
+
+  private static final String CRT_HEADER = "X-CHAP";
+  private static final String AUTH_URI = "https://helios/_auth";
+
+  private final RequestDispatcher dispatcher;
+  private final String username;
+  private final CrtAuthClient crtClient;
+  private volatile String token;
+  private final Object lock = new Object();
+
+  public CrtAuthProvider(final Context context,
+                         final String authServer) {
+    this.dispatcher = context.dispatcher();
+    this.username = context.user();
+    try {
+      this.crtClient = new CrtAuthClient(new AgentSigner(), authServer);
+    } catch (CrtAuthException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @VisibleForTesting
+  CrtAuthProvider(final RequestDispatcher dispatcher,
+                  final String authServer,
+                  final String username,
+                  final Signer signer) {
+    this.dispatcher = dispatcher;
+    this.username = username;
+    this.crtClient = new CrtAuthClient(signer, authServer);
+  }
+
+  @Override
+  public String currentAuthorizationHeader() {
+    // TODO (dxia) Should we use a lock here? It doesn't really matter if the token isn't
+    // thread-safe. We should get valid tokens each time. The lock just prevents
+    // another thread from doing extra work to get an extra token when it could just wait.
+    if (token == null) {
+      synchronized (lock) {
+        try {
+          token = renewAuthorizationHeader(null).get();
+        } catch (InterruptedException | ExecutionException e) {
+          log.error("Failed to complete crtauth handshake.");
+          throw Throwables.propagate(e);
+        }
+      }
+    }
+
+    return token;
+  }
+
+  @Override
+  public ListenableFuture<String> renewAuthorizationHeader(final String ignored) {
+    final String authRequest = CrtAuthClient.createRequest(username);
+    try {
+      final HeliosRequest request = HeliosRequest.builder()
+          .method("GET")
+          .uri(new URI(AUTH_URI))
+          .appendHeader(CRT_HEADER, "request:" + authRequest)
+          .build();
+
+      final ListenableFuture<Response> challengeFuture = dispatcher.request(request);
+
+      final ListenableFuture<Response> tokenFuture =
+          Futures.transform(challengeFuture, new AsyncFunction<Response, Response>() {
+            @Override
+            public ListenableFuture<Response> apply(@NotNull final Response response)
+                throws Exception {
+              final String challengeHeader = response.header(CRT_HEADER);
+              // TODO (dxia) Should we check status codes or just simply check headers as I've done?
```

Comment:
should check status code, I believe the reference python implementation does (and the server will indeed send 200 OK as long as everything is OK)